File: code/py/src/luogu/contest/ioi20230730/QC.py
def combinationSum2(nums, target):
    nums.sort()  # 首先对数组进行排序，方便后面的去重处理
    res = []  # 用于存储符合条件的组合

    def backtrack(start, target, path):
        if target == 0:  # 如果目标数为0，说明找到了一个符合条件的组合
            res.append(path)
            return
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i - 1]:  # 去除重复的情况
                continue
            if target < nums[i]:  # 剪枝：如果当前数字大于目标数，后面的数字肯定也大于，直接结束循环
                break
            # 递归搜索下一层
            backtrack(i + 1, target - nums[i], path + [nums[i]])

    backtrack(0, target, [])
    return len(res)

# 读取输入
n, target = map(int, input().split())
nums = list(map(int, input().split()))

# 调用函数并输出结果
print(combinationSum2(nums, target))

File: code/java/src/onlinejudge/luogu/contest/c20230730njit/QC.java
package luogu.contest.c20230730njit;

import java.util.*;

class QC {
    static int n, target, numsSize;
    static List<Integer> nums;
    static int res = 0;

    static void input() {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        target = scanner.nextInt();
        nums = new ArrayList<>(n);
        int tmp;
        for (int i = 0; i < n; i++) {
            tmp = scanner.nextInt();
            if (tmp < target) {
                nums.add(tmp);
            } else if (tmp == target) {
                ++res;
            }
        }
        scanner.close();
        nums.sort(Comparator.comparingInt(x -> x));
        numsSize = nums.size();
    }

    static void dfs(int sum, int start) {
        if (target == sum) {
            ++res;
        } else if (target > sum) {
            int last = -1;
            for (int i = start + 1; i < numsSize; i++) {
                if (last != nums.get(i)) {
                    dfs(sum + nums.get(i), i);
                    last = nums.get(i);
                }
            }
        }
    }

    public static void main(String[] args) {
        input();
        Deque<int[]> queue;
        int[] vertex;
        int last = -1, lastDfs;
        for (int i = 0; i < numsSize; i++) {
            if (last != nums.get(i)) {
//                dfs(nums.get(i), i);

                queue = new ArrayDeque<>(n);
                queue.push(new int[]{nums.get(i), i});
                while (!queue.isEmpty()) {
                    vertex = queue.pop();
                    if (target == vertex[0]) {
                        ++res;
                    } else if (target > vertex[0]) {
                        lastDfs = -1;
                        for (int j = vertex[1] + 1; j < numsSize; j++) {
                            if (lastDfs != nums.get(j)) {
                                queue.push(new int[]{vertex[0] + nums.get(j), j});
                                lastDfs = nums.get(j);
                            }
                        }
                    }
                }

                last = nums.get(i);
            }
        }
        System.out.println(res);
    }
}